### PR TITLE
SageMaker: Fix sagemaker DDP & metric logs

### PR DESCRIPTION
**Pull request by philschmid**
# What does this PR do?

This PR fixes the fix introduced in #12853. Since `sm_dist.Barrier()` is not available in `smd 1.0.0 2021-01-26` release and smd `1.2.0`, which are used for the DLC with PyTorch 1.7.1 & 1.8.1 (they maintained ones). Further is the fix `sm_dist.barrier()` also working with `smd 1.0.0 2020-12-06`.  
cc @sgugger 

Additionally, does this PR update:
* the SageMaker test image_uris
* instances type for distributed training -> there are capacity issues with the 24xlarge
* and moves the adding of the `StreamHandler(sys.stdout)` for logs to the `trainer_pt_utils.py` to also cover the `log_metrics function` more to this below.

---
When running a training job on SageMaker all stdout or stderr are sent to Amazon CloudWatch Logs. With the introduction of the new `log_metrics` function, SageMaker lost its output. Therefore I moved the `StreamHandler(sys.stdout)` to the `trainer_pt_utils.py` and removed it in the `trainer`. More information can be found here #10633

### CODE REVIEW

File: tests/sagemaker/scripts/pytorch/requirements.txt [hunk 1]
```diff
@@ -1 +1,2 @@
-git+https://github.com/huggingface/transformers.git@master # install master or adjust ist with vX.X.X for installing version specific transforms
\ No newline at end of file
+git+https://github.com/huggingface/transformers.git@master # install master or adjust ist with vX.X.X for installing version specific transforms
```
Comment (stas00) [inline tests/sagemaker/scripts/pytorch/requirements.txt:1]:
```suggestion
git+https://github.com/huggingface/transformers.git@master # install master or adjust it with vX.X.X for installing version specific transforms
```

**Review by stas00 — APPROVED** [review]
LGTM

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @philschmid

**Comment by sgugger** [comment]
Thanks a lot @philschmid !

**Pull request merged**